Macros/Macros.js

```diff
@@ -192,6 +223,64 @@ class FurnaceMacros {
         div.find(".furnace-macro-expand").on('click', (ev) => div.toggleClass("fullscreen"));
         refreshHighlight();
     }
+
+    static async _execute(message) {
+        let macro = game.macros.getName(message.macroName);
+        if (!macro || !getProperty(macro.data, "flags.furnace.gmexecute" || !game.users.get(macro.data.author).isGM)) {
+            ui.notification.error(`User ${message.userId} tried to invailidly GM execute macro ${message.macroName}`);
+            return "Invalid Request";
+        }
+
+        try {
+            return await macro.renderContent("RemoteContext", message, ...message.args);
+        } catch (err) {
+            ui.notifications.error(`There was an error in your macro syntax. See the console (F12) for details`);
+            console.error(err);
+            return "Macro syntax error";
+        }
+    }
+
+    // request remote execution of a macro
+    static async requestRemoteExecute(macroName, ...args) {
+        let targetGM = game.users.entities.find(u => u.isGM && u.active);
+        let macroMessage = {
+            action: "GMExecuteMacro", 
+            requestId: "remoteResult" + game.user.id + Date.now(), 
+            userId: game.user.id, 
+            targetGM: targetGM._id, 
+            macroName, 
+            sceneId: canvas.scene.id, // pass this so macros can know which scene the player is on.
+            tokenId: ChatMessage.getSpeaker().token, 
+            event: event || {event: {shiftKey: true}}, // this is to keep minor-qol happy
+            args
+        };
+        if (game.user.isGM) return await FurnaceMacros._execute(macroMessage); 
+        // setup to handle exectuion response
+        let requestResponse = await new Promise((resolve, reject) => {
+            Hooks.once(macroMessage.requestId, resolve);
+            game.socket.emit(FurnaceMacros._moduleSocket, macroMessage)
+            setTimeout(() => Hooks.callAll(macroMessage.requestId, "timeout"), 5000);
+        })
+       
+        if (requestResponse === "timeout") return "Request Timeout";
+        else return requestResponse.result;
+    }
+
+    _setupSocket() {
+        Hooks.once("ready", () => {
+            game.socket.on(FurnaceMacros._moduleSocket, async (message) => {
+                if (message.action === "GMExecuteMacro") {
+                    if (game.user.id !== message.targetGM) return;
```

Comment:
One issue I see is if the GM has more than one window open (popout or just multiple tabs open). It will cause a macro to get executed more than once, which may be problematic.